Add `all` configuration

darkred opened the pull request:
Implements #247 

I hope I didn't miss anything.


## Note to reviewers:

I see that the last test action `CI / update` [fails](https://github.com/ota-meshi/eslint-plugin-regexp/pull/250/checks?check_run_id=2819106682) but I can't find why.

ota-meshi commented:
> I see that the last test action CI / update fails but I can't find why.

Run the `npm run update` command.

darkred commented:
I just did (while on my `add-all-configuration` branch)
and the resulting uncommitted changes reverted my changes in `README.md` and `all.ts` *(it's now identical to `recommended.ts`).
Please see screenshots of the changes: [1](https://user-images.githubusercontent.com/723651/121894995-70377280-cd28-11eb-91cc-4032bac3c069.jpg), [2](https://user-images.githubusercontent.com/723651/121895007-72013600-cd28-11eb-9b0d-6e9fcad308e2.jpg), [3](https://user-images.githubusercontent.com/723651/121895021-74fc2680-cd28-11eb-96ed-08f0d519f436.jpg)

RunDevelopment commented on docs/user-guide/README.md:
````diff
@@ -18,13 +18,18 @@ npm install --save-dev eslint eslint-plugin-regexp
+The `plugin:regexp/all` config enables all rules. It's meant for testing, not for production use because it changes with every minor and major version of the plugin. Use it at your own risk.
+  See [https://github.com/ota-meshi/eslint-plugin-regexp/blob/master/lib/configs/all.ts](https://github.com/ota-meshi/eslint-plugin-regexp/blob/master/lib/configs/all.ts) for more details.*
+
 ```js
````
This code block should be in the section of the recommended configuration.

I don't think we need a code example for this config. We already have this example for the recommended config and people should be able to figure out which word to replace with `all`.

RunDevelopment commented on docs/user-guide/README.md:
```diff
@@ -18,13 +18,18 @@ npm install --save-dev eslint eslint-plugin-regexp
+### The all configuration
+
+The `plugin:regexp/all` config enables all rules. It's meant for testing, not for production use because it changes with every minor and major version of the plugin. Use it at your own risk.
+  See [https://github.com/ota-meshi/eslint-plugin-regexp/blob/master/lib/configs/all.ts](https://github.com/ota-meshi/eslint-plugin-regexp/blob/master/lib/configs/all.ts) for more details.*
```
```suggestion
  *See [https://github.com/ota-meshi/eslint-plugin-regexp/blob/master/lib/configs/all.ts](https://github.com/ota-meshi/eslint-plugin-regexp/blob/master/lib/configs/all.ts) for more details.*
```

RunDevelopment commented on docs/user-guide/README.md:
```diff
@@ -18,13 +18,18 @@ npm install --save-dev eslint eslint-plugin-regexp
 
 Add `regexp` to the plugins section of your `.eslintrc` configuration file (you can omit the `eslint-plugin-` prefix)  
-and either use the recommended configuration or configure the rules you want:
+and either use one of the two configurations available (`recommended` and `all` ) or configure the rules you want:
```
```suggestion
and either use one of the two configurations available (`recommended` or `all`) or configure the rules you want:
```

RunDevelopment commented:
I think the easiest way to make the all config is to import the recommended config, make a copy of it and then add all rules that aren't included already. This would also mean that won't have to change any `scripts/*` files and won't have problems with `npm run update`.

darkred commented:
@ota-meshi 
To be sure, I deleted my cloned directory altogether, then cloned again my fork.

Then I just updated README.md (I mean the [root](https://github.com/ota-meshi/eslint-plugin-regexp/blob/master/README.md) one)
and ran `npm run update`
and now my `lib\configs\all.ts` once again became identical to `lib\configs\recommended.ts` (uncommitted changes).

How can I fix this? What have I missed?

RunDevelopment commented:
> now my `lib\configs\all.ts` once again became identical to `lib\configs\recommended.ts` (uncommitted changes).
> 
> How can I fix this? What have I missed?

Well, in `tools/update-rulesets.ts` you instruct `npm run update` to do exactly that. Just revert the file.

RunDevelopment commented on lib/configs/all.ts:
```diff
@@ -0,0 +1,71 @@
+export = {
+    plugins: ["regexp"],
+    rules: {
+        // ESLint core rules
+        "no-control-regex": "error",
```
I still think that it would be better to generate this config at runtime.

We might have to disable/configure more ESLint core rules in the future and I don't want this config to become out of date.

ota-meshi replied:
I think the config of `eslint:all` will be helpful for you.
https://github.com/eslint/eslint/blob/master/conf/eslint-all.js

You can get all the rules from the following files.
https://github.com/ota-meshi/eslint-plugin-regexp/blob/master/lib/utils/rules.ts

You can access the rule name from `rule.meta.docs.ruleId`.
https://github.com/ota-meshi/eslint-plugin-regexp/blob/master/lib/types.ts#L22
https://github.com/ota-meshi/eslint-plugin-regexp/blob/master/lib/utils/index.ts#L177

darkred replied:
And, I was wondering how to make a this `all.ts` always include every single plugin's rule.
Maybe it should be automatically generated from `recommended.ts` with concatenating parts/lines of `lib\utils\rules.ts` ?

Because I made this `all.ts` based on `recommended.ts` :
I ran `dir /b lib\rules > rules.txt` in command prompt, 
removed the `.ts` extension from all rows,
and then appended `"regexp/` and `": "error",` in each row start/end.

&nbsp;
So, is my PR is not helpful? Should I close it? Just asking.

RunDevelopment replied:
No, don't worry, it is helpful. We just have to change `all.ts` a little.

I wouldn't want `all.ts` to be generated like `recommended.ts` because it adds to the complexity of our tooling and isn't necessary.

I was imagining something like this:

```ts
import recommended from "./recommended"
import { rules } from "../utils/rules"

const all: Record<string, string> = {}
for (const rule of rules) {
    all[rule.meta.docs.ruleId] = "error"
}

export = {
    plugins: ["regexp"],
    rules: {
        ...all,
        ...recommended.rules,
    },
}
```

I haven't tested it though.

darkred replied:
I just committed this, as it seems to be fine.

---

<details>
<summary>Btw, it's maybe a bit off-topic but:</summary>
It's the 2nd PR I made in the repo, 
and I feel once again embarrassed being confused over the same things,
i.e. what README.md should I modify vs which are automatically generated + the usage of `npm run update`.
I feel bad asking for help every time.

So, I'd like to kindly ask for something, please:
could someone make a dedicated CONTRIBUTING.md, clarifying in details: 
- how to make PR for the documentation, 
- how to make PR for the code itself (proposing a new rule, etc).
- how to test the plugin afterwards (I mean how to configure my .eslintrc.js to use the produced `dist` directory)

Regarding 1st and 2nd, I mean, which files to modify, which are to be left intact, 
and don't forget to `npm run update` afterwards.

&nbsp;
The info in README.md is helpful, but not enough:
![2021-06-14_164741](https://user-images.githubusercontent.com/723651/121902797-5437cf00-cd30-11eb-99ae-b4ea1530cd68.jpg)

&nbsp;
I believe it would be very helpful, and facilitate new contributors.

</details>

RunDevelopment replied:
> could someone make a dedicated CONTRIBUTING.md guide

Good point, I'll write one.

RunDevelopment commented on README.md:
```diff
@@ -46,13 +46,16 @@ npm install --save-dev eslint eslint-plugin-regexp
 The `plugin:regexp/recommended` config enables a subset of [the rules](#white_check_mark-rules) that should be most useful to most users.
 *See [https://github.com/ota-meshi/eslint-plugin-regexp/blob/master/lib/configs/recommended.ts](https://github.com/ota-meshi/eslint-plugin-regexp/blob/master/lib/configs/recommended.ts) for more details.*
 
+The `plugin:regexp/all` config enables all rules. It's meant for testing, not for production use because it changes with every minor and major version of the plugin. Use it at your own risk.
```
Last nit from my side: I think this should get its own section after the "Advanced Configuration" section. ESLint doc also has a similar structure.

I would probably call the section "Using `plugin:regexp/all`".

darkred replied:
Done.

ota-meshi reviewed (approved):
LGTM! Thank you very much for your contribution!

darkred commented:
You are welcome. I'm glad I could contribute!
Thank you both for the help.
